marimo/_save/hash.py

```diff
@@ -269,76 +321,209 @@ def __init__(
         self.cache_type = cache_type
         self.hash_alg.update(hash_raw_module(module, hash_type))
-        self.hash = (
-            base64.urlsafe_b64encode(self.hash_alg.digest())
-            .decode("utf-8")
-            .strip("=")
+
+    @staticmethod
+    def from_parent(
+        parent: BlockHasher,
+    ) -> BlockHasher:
+        # Use a previous block as the basis of a new block.
+        block = BlockHasher.__new__(BlockHasher)
+        block.module = parent.module
+        block.graph = parent.graph
+        block.cell_id = parent.cell_id
+        block.pin_modules = parent.pin_modules
+        block.fn_cache = {}
+        if parent.fn_cache is not None:
+            block.fn_cache = dict(parent.fn_cache)
+        block.visitor = parent.visitor
+        block.defs = set(parent.defs)
+        block.stateful_refs = set(parent.stateful_refs)
+        block.hash_alg = parent.hash_alg.copy()
+        block._hash = None
+        block.cache_type = parent.cache_type
+        block.content_refs = set(parent.content_refs)
+        block.execution_refs = set(parent.execution_refs)
+        block.context_refs = set(parent.context_refs)
+        return block
+
+    @property
+    def hash(self) -> str:
+        if self._hash is None:
+            assert self.hash_alg is not None, "Hash algorithm not initialized."
+            self._hash = (
+                base64.urlsafe_b64encode(self.hash_alg.digest())
+                .decode("utf-8")
+                .strip("=")
+            )
+        return self._hash
+
+    def __hash__(self) -> int:
+        return hash(self.hash)
+
+    def _apply_content_hash(
+        self, content_serialization: dict[Name, bytes]
+    ) -> None:
+        self._hash = None
+        for ref in sorted(content_serialization):
+            self.hash_alg.update(content_serialization[ref])
+
+    def collect_for_content_hash(
+        self,
+        refs: set[Name],
+        scope: dict[str, Any],
+        ctx: Optional[RuntimeContext],
+        scoped_refs: set[Name],
+        apply_hash: bool = True,
+    ) -> SerialRefs:
+        self._hash = None
+        refs, content_serialization, _ = (
+            self.serialize_and_dequeue_content_refs(refs, scope)
         )
+        # If scoped refs are present, then they are unhashable
+        # and we should fallback to normal hash or fail.
+        if unhashable := (refs & scoped_refs) - self.execution_refs:
+            # pickle is a python default
+            import pickle
+
+            failed = []
+            exceptions = []
+            # By rights, could just fail here - but this final attempt should
+            # provide better user experience.
+            for ref in unhashable:
+                try:
+                    _hashed = pickle.dumps(scope[ref])
+                    content_serialization[ref] = type_sign(_hashed, "pickle")
+                    refs.remove(ref)
+                except (pickle.picklingError, TypeError) as e:
```

Comment:
```suggestion
                except (pickle.PicklingError, TypeError) as e:
```

my b, had a typo in my original suggestion